src/core/matcher.rkt

```diff
@@ -115,25 +116,27 @@
            (display ">" port))])
 
 (define (rewrite-expression expr #:destruct [destruct? #f] #:root [root-loc '()])
+  (define env (for/hash ([v (free-variables expr)]) (values v 'real)))
   (reap [sow]
-    (for ([rule (*rules*)])
+    (for ([rule (if (equal? 'complex (type-of expr env)) (*complex-rules*) (*rules*))])
       (let* ([applyer (if destruct? rule-apply-force-destructs rule-apply)]
              [result (applyer rule expr)])
         (when result
             (sow (change rule root-loc (cdr result))))))))
 
 (define (rewrite-expression-head expr #:root [root-loc '()] #:depth [depth 1])
 
+  (define env (for/hash ([v (free-variables expr)]) (values v 'real)))
   (define (rewriter expr ghead glen loc cdepth)
     ; expr _ _ _ _ -> (list (list change))
     (reap (sow)
-          (for ([rule (*rules*)])
-            (when (or
-                    (not ghead) ; Any results work for me
-                    (and
-                      (list? (rule-output rule))
-                      (= (length (rule-output rule)) glen)
-                      (eq? (car (rule-output rule)) ghead)))
+          (for ([rule (if (equal? 'complex (type-of expr env)) (*complex-rules*) (*rules*))])
+              (when (or
```

Comment:
In the future, it's better to avoid this sort of whitespace-only change, since it makes it harder to track down what sorts of changes are made.